molgenis-data/src/main/java/org/molgenis/data/decorator/PermissionCheckingDecorator.java

```diff
@@ -0,0 +1,188 @@
+package org.molgenis.data.decorator;
+
+import org.molgenis.data.*;
+import org.molgenis.data.support.QueryImpl;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * Decorator that checks if actions on this repository are allowed using a PermissionChecker.
+ *
+ * @param <E> the class of the entities in the repository
+ */
+public class PermissionCheckingDecorator<E extends Entity> extends AbstractRepositoryDecorator<E>
+{
+	private PermissionChecker<E> permissionChecker;
+
+	public PermissionCheckingDecorator(Repository<E> delegateRepository, PermissionChecker<E> permissionChecker)
+	{
+		super(delegateRepository);
+		this.permissionChecker = requireNonNull(permissionChecker);
+	}
+
+	@Override
+	public Iterator<E> iterator()
+	{
+		Iterable<E> iterable = () -> delegate().iterator();
+		return stream(iterable.spliterator(), false).filter(
+				entity -> permissionChecker.isReadAllowed(entity.getIdValue())).iterator();
```

Comment:
How about adding a default method in PermissionChecker boolean `isReadAllowed(E entity)` that calls isReadAllowed(Object entityId)?